Poker game step 3

BMO5 opened the pull request:
**지난 PR문제점**
- 이전 Step에서 PR, 머지 후 피드백을 반영하기 위해 브랜치를 옮기지 않고 작업을 했습니다.
- 작업 후 제 브랜치로 돌아가 upstream과 싱크를 맞추고 피드백을 반영한 브랜치를 다시 머지했습니다.
- 그러자 피드백 이전(upstream에 머지가 이미 된)의 커밋로그까지 같이 머지되어 커밋이 중복되었습니다.
- 그래서 다음 Step 완료 후 PR을 보내는데 이전의 이미 머지된 커밋로그까지 몽땅 추가되었습니다.

**해결 방법**
- upstream의 위치에서 새로운 브랜치를 만들고, 이후의 작업들 중 중복되지 않은 커밋을 `git cherry-pick 커밋명`으로 복사했습니다. 
- 일일이 커밋명을 찾아 복사하기 보다 `git cherry-pick oldest-commit^..latest-commit`을 사용해 특정 범위의 커밋을 복사했습니다.
- 새로운 브랜치를 푸시 후 PR보냅니다.

**질문**   
- 마스터가 머지해주시고 피드백을 남겨주시는데, 이 피드백을 반영할때도 새로운 브랜치를 만들어서 작업을 하는게 좋을까요?

알게된 점
- 구조체와 클래스의 가장 큰 차이점을 꼽자고 한다면 상속과 값/참조 타입인 것 같습니다.
- shuffle 알고리즘 중 fisher-yates 알고리즘을 찾아보았습니다. 이는 예전버전(O(N^2))과 현대버전(O(N))이 있는데 Swift의 shuffle()은 현대버전을 사용합니다.
- deinit을 통해 클래스의 메모리관리를 알아보았습니다. 카드를 그냥 뽑을 경우 메모리가 해제되지만, 새로운 프로퍼티에 반환하면 이 프로퍼티가 nil이 되어야 해제가 되었습니다. reset시에는 남아있는 메모리를 모두 해제한 후 다시 할당하는 것을 확인할 수 있었습니다.

BMO5 committed:
피드백 반영

CustomStringConvertible을 채택해 cardInfo() -> String 메소드를 대체함
rank의 rawValue를 다시 1~13으로 설정.

BMO5 committed:
피드백 반영 - 주석 제거, 뷰 컨트롤러 print 수정

로직을 설명하는 불필요한 주석 제거
뷰 컨트롤러에서 PokerCard를 출력하는 방법 수정

BMO5 committed:
CardDeck 구조체 생성

BMO5 committed:
PokerCard 내부 enum에 CaseIterable 추가

카드덱을 만들 때 열거형 항목 순회를 하기위해

BMO5 committed:
CardDeck 구조체 생성, 카드 객체 인스턴스 구현

이니셜라이저로 구조체의 인스턴스가 생성될 때 카드 객체 인스턴스 생성

BMO5 committed:
count 구현

BMO5 committed:
shuffle() 구현

BMO5 committed:
removeOne(), reset() 구현

BMO5 committed:
Test 코드 작성

godrm commented:
> **해결 방법**
> 
> * upstream의 위치에서 새로운 브랜치를 만들고, 이후의 작업들 중 중복되지 않은 커밋을 `git cherry-pick 커밋명`으로 복사했습니다.
> * 일일이 커밋명을 찾아 복사하기 보다 `git cherry-pick oldest-commit^..latest-commit`을 사용해 특정 범위의 커밋을 복사했습니다.
> * 새로운 브랜치를 푸시 후 PR보냅니다.
> 
> **질문**
> 
> * 마스터가 머지해주시고 피드백을 남겨주시는데, 이 피드백을 반영할때도 새로운 브랜치를 만들어서 작업을 하는게 좋을까요?

잘하셨습니다. cherry-pick 세부 옵션까지 공유해줘서 좋습니다. 항상 다른 방법이 있으니까 해결하는 게 좋습니다.
피드백에 대해서는 PR 오픈해놓고 수정하는 경우도 있고, 아닌 경우도 있습니다. 
전자는 기존 브랜치에 계속 PUSH하면 되고 후자는 새 브랜치에서 작업하고 커밋해놓고, 같은 브랜치에서 다음 단계를 진행해도 됩니다.

godrm commented on PokerGameApp/PokerGameApp/CardDeck.swift:
```diff
@@ -0,0 +1,48 @@
+                tempCards.append(PokerCard(suit: suit, rank: rank))
+            }
+        }
+        return tempCards
```
어차피 내부함수라면 return 하는게 아니라 self.Cards에 직접 저장해도 되지 않을까요?

BMO5 replied:
그러네요! init에 넣으려고 return을 해준건데, self.cards에 직접 저장해도 init에 호출할 수 있네요. 수정해보겠습니다!

godrm commented on PokerGameApp/PokerGameApp/CardDeck.swift:
```diff
@@ -0,0 +1,48 @@
+import Foundation
+
+struct CardDeck {
+    var Cards: [PokerCard] = []
```
타입을 선언할 때는 대문자로 시작하고, 반대로 속성이나 변수를 선언할 때는 소문자로 해야 헷갈리지 않습니다.

BMO5 replied:
앗 ..! 저거 변수명인데 왜 대문자지 .. 바로 수정하겠습니다.

godrm commented on PokerGameApp/PokerGameApp/CardDeck.swift:
```diff
@@ -0,0 +1,48 @@
+    }
+    
+    mutating func reset() {
+        self.Cards = makeCards()
```
return 하도록 하니까 이렇게 한꺼번에 교체할 수 있는 장점은 있네요.

godrm commented on PokerGameApp/PokerGameApp/PokerCard.swift:
```diff
@@ -7,33 +7,42 @@
+        default:
+            return "\(self.suit.rawValue)\(self.rank.rawValue)"
+        }
+        return "\(self.suit.rawValue)\(over10)"
```
Card 타입이 CustomStringConvertible를 채택한 것처럼 
Suit 나 Rank도 CustomStringConvertible를 채택해보세요. 그러면 여기 코드는 더 간결해질 겁니다.

godrm commented on PokerGameApp/PokerGameApp/PokerCard.swift:
```diff
@@ -7,33 +7,42 @@
+        case jack, queen, king
     }
     
     let suit:Suit, rank:Rank
```
서로 다른 타입인데 한 줄로 쓰기 보다는 별도로 쓰는 게 좋겠습니다.

BMO5 replied:
넵. 감사합니다. 피드백주신 부분 수정 후 푸시했습니다!

godrm reviewed (changes requested):
PR이 깔끔하네요. 아래 내용 확인해보고 PR 열어놓을테니 CustomStringConvertible 프로토콜 부분만 개선해서 다시 PUSH해보세요.

BMO5 committed:
변수명 Cards -> cards로 수정

rename을 사용하니 깔끔히 모두 수정됨

BMO5 committed:
makeCards() -> reset() 으로 변경

내부함수라 return하기보다 self.cards에 직접 저장했다.
그러니 makeCards() 메소드와 이전에 있던 reset() 메소드가 동일하여 하나로 통일했다.

BMO5 committed:
reset()의 private 제거

BMO5 committed:
CustomStringConvertible 프로토콜 개선

클래스에 채택해도 되지만, 그걸 직접 사용하는 곳에 프로토콜을 채택하자
다른 부분의 코드가 간결해졌다

BMO5 committed:
README에 Step 2,3 추가

BMO5 commented:
말씀해주신 부분 개선해서 PUSH 했습니다!

godrm commented on PokerGameApp/PokerGameApp/CardDeck.swift:
```diff
@@ -8,23 +8,23 @@
 
 struct CardDeck {
-    var Cards: [PokerCard] = []
+    var cards: [PokerCard] = []
```
내부 속성은 private var를 우선적으로 고려하세요.

BMO5 replied:
넵. 알겠습니다. 다음 스텝에서 수정하겠습니다!
